spacy/tokens/doc.pyx

```diff
@@ -1451,6 +1451,70 @@ cdef class Doc:
                 remove_label_if_necessary(attributes[i])
                 retokenizer.merge(span, attributes[i])
 
+    @staticmethod
+    def from_json(vocab, doc_json):
+        """Convert a JSON document generated by Doc.to_json() to a Doc.
+
+        vocab (Vocab): A storage container for lexical types.
+        doc_json (dict): The Doc data in JSON format.
+        RETURNS (Doc): A doc instance corresponding to the specified JSON representation.
+        """
+
+        words = []
+        token_annotations = {
+            # For each annotation type: store (1) annotation type required to include it, (2) list of annotations for
+            # this type.
+            annotation_type: {"req": annotation_type if annotation_type != "head" else "dep", "values": None}
+            for annotation_type in ("tag", "pos", "morph", "lemma", "dep", "head")
+        }
+
+        # Gather token-level properties.
+        for token in doc_json["tokens"]:
+            words.append(doc_json["text"][token["start"]: token["end"]])
+            for annotation_type in token_annotations:
+                # Assuming all tokens have exactly the same set of attributes.
+                if token_annotations[annotation_type]["req"] in token:
+                    if token_annotations[annotation_type]["values"] is None:
+                        token_annotations[annotation_type]["values"] = []
+                    token_annotations[annotation_type]["values"].append(token[annotation_type])
+
+        # Create Doc instance.
+        doc = Doc(
+            vocab,
+            words=words,
+            tags=token_annotations["tag"]["values"],
+            pos=token_annotations["pos"]["values"],
+            morphs=token_annotations["morph"]["values"],
+            lemmas=token_annotations["lemma"]["values"],
+            heads=token_annotations["head"]["values"],
+            deps=token_annotations["dep"]["values"],
+        )
```

Comment:
You'll definitely need to define `spaces` as well - this stores whether or not a space should follow a word. You can create a failing unit test that shows the current behaviour is not correct:

Take a sentence `"This is just brilliant."`. After `to_json` and `from_json` it will currently become `"This is just brilliant ."`

Reply:
Yes, as this is an instance method now, spaces are set differently. A unit test for space preservation will be added in the next commit.